server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -76,28 +82,41 @@ public IndexInput fetchBlob(BlobFetchRequest blobFetchRequest) throws IOExceptio
     }
 
     @SuppressWarnings("removal")
-    private static FileCachedIndexInput createIndexInput(FileCache fileCache, BlobContainer blobContainer, BlobFetchRequest request) {
+    private static FileCachedIndexInput createIndexInput(
+        FileCache fileCache,
+        BlobContainer blobContainer,
+        List<BlobFetchRequest> requestList
+    ) {
         // We need to do a privileged action here in order to fetch from remote
         // and write to the local file cache in case this is invoked as a side
         // effect of a plugin (such as a scripted search) that doesn't have the
         // necessary permissions.
+        assert requestList.isEmpty() == false;
         return AccessController.doPrivileged((PrivilegedAction<FileCachedIndexInput>) () -> {
             try {
-                if (Files.exists(request.getFilePath()) == false) {
-                    try (
-                        InputStream snapshotFileInputStream = blobContainer.readBlob(
-                            request.getBlobName(),
-                            request.getPosition(),
-                            request.getLength()
-                        );
-                        OutputStream fileOutputStream = Files.newOutputStream(request.getFilePath());
-                        OutputStream localFileOutputStream = new BufferedOutputStream(fileOutputStream)
-                    ) {
-                        snapshotFileInputStream.transferTo(localFileOutputStream);
+                boolean needsAppend = false;
+                if (Files.exists(requestList.get(0).getFilePath()) == false) {
+                    for (BlobFetchRequest request : requestList) {
+                        try (
+                            InputStream snapshotFileInputStream = blobContainer.readBlob(
+                                request.getBlobName(),
+                                request.getPosition(),
+                                request.getLength()
+                            );
+                            OutputStream fileOutputStream = needsAppend
+                                ? Files.newOutputStream(request.getFilePath(), APPEND)
```

Comment:
I think you should only need to open the local file once, i.e. create the local file and then start looping over the remote parts writing to the local file. Then the local file gets closed after all parts have been downloaded.

Reply:
Without using APPEND, every loop we will be overriding the local file, so always only last chunk will be written. I have updated this logic in latest commit, instead of create a new outputstream everytime create it only twice: once for creating empty file, another for specifying append option. (without having a new file already created append option will fail).

Please let me know if you have any other suggestions here